src/ripple/app/misc/SHAMapStoreImp.cpp

```diff
@@ -666,16 +705,32 @@ SHAMapStoreImp::health()
     }
     if (!netOPs_)
         return Health::ok;
+    assert(deleteInterval_);
 
-    constexpr static std::chrono::seconds age_threshold(60);
-    auto age = ledgerMaster_->getValidatedLedgerAge();
-    OperatingMode mode = netOPs_->getOperatingMode();
-    if (mode != OperatingMode::FULL || age > age_threshold)
+    if (healthy_)
     {
-        JLOG(journal_.warn()) << "Not deleting. state: "
-                              << app_.getOPs().strOperatingMode(mode, false)
-                              << ". age " << age.count() << 's';
-        healthy_ = false;
+        auto age = ledgerMaster_->getValidatedLedgerAge();
+        OperatingMode mode = netOPs_->getOperatingMode();
+        if (recoveryBuffer_ && mode == OperatingMode::SYNCING &&
+            age < ageThreshold_)
+        {
+            JLOG(journal_.warn())
+                << "Waiting " << recoveryBuffer_->count()
+                << "s for node to get back into sync with network. state: "
+                << app_.getOPs().strOperatingMode(mode, false) << ". age "
+                << age.count() << 's';
+            std::this_thread::sleep_for(*recoveryBuffer_);
+
+            age = ledgerMaster_->getValidatedLedgerAge();
+            mode = netOPs_->getOperatingMode();
+        }
```

Comment:
Notwithstanding recovery_buffer_seconds, the online_delete activity is triggered each time a new ledger is validated. And desyncs seem to usually happen for an entire consensus interval, of ~3-4 seconds. This really seems superfluous, especially adding a new config option to tune this.

Reply:
I'm not married to it, but on my local node, it was taking so long for the ledger to copy that the node would switch to `syncing` before it could finish, and thus it would abort only to almost immediately start rotating again. The thing is that as far as I could tell, it wasn't losing sync because of the ledger copy, but because it build a different ledger during consensus than the UNL validators. Here's an example that includes some log messages that I added only to figure out what was going on:

```
2020-May-27 19:55:17.7525561 UTC SHAMapStore:TRC Nothing to delete from DELETE FROM AccountTransactions WHERE LedgerSeq < %u;
2020-May-27 19:55:17.7547395 UTC SHAMapStore:TRC copying ledger 55752529
2020-May-27 19:55:35.0709935 UTC NetworkOPs:NFO consensusViewChange
2020-May-27 19:55:35.0718717 UTC NetworkOPs:NFO STATE->syncing
2020-May-27 19:55:35.1798535 UTC SHAMapStore:WRN Not deleting. state: syncing. age 1s
2020-May-27 19:55:35.1803785 UTC SHAMapStore:WRN Not deleting. state: syncing. age 1s
2020-May-27 19:55:37.7108585 UTC SHAMapStore:WRN Not deleting. state: syncing. age 0s
2020-May-27 19:55:38.6848155 UTC NetworkOPs:NFO STATE->tracking
2020-May-27 19:55:38.6850136 UTC NetworkOPs:NFO STATE->full
2020-May-27 19:55:41.6406001 UTC SHAMapStore:WRN rotating  validatedSeq 55752535 lastRotated_ 55461357 deleteInterval 30000 canDelete_ 55461356 state full age 0s
```
Note the timestamps. At 19:55:17, it starts copying the ledger. 28 seconds later, the consensus view changes and it drops to syncing. 0.1 seconds later, the ledger copy aborts in `health()`. About 3 seconds after that (the next ledger, I presume), we're back to tracking, then full. 3 seconds later, we start rotating again. My node was doing this for _hours_. It's a beefy machine, but a Windows debug build.

The reason I added it as a config option is so that it _wouldn't_ do this by default. 

The obvious question is whether it's worth adding a feature to support marginal nodes. It seemed like a good idea at the time, and it at least helped me get this feature developed and finished.

I can think of two options:
* Remove it entirely.
* Make the feature `#if DEBUG`, remove the config option, and use a hard-coded value.

I'll go with whatever you think makes more sense.

Reply:
It could be useful in a troubleshooting context--since you've already done the work, may as well keep it, as an option, but not by default. It's not plausible from that vantage point to determine whether the node itself is flaky, or if online_delete triggered the desync. So, to be safe, assume the latter, which means the default behavior should be to not have a buffer of time to wait for the node to recover.

Reply:
To confirm, the default time is to not have the buffer by default. :+1: